igneous/tasks.py

```diff
@@ -943,3 +943,230 @@ def _get_map(self):
+        
+        assert np.any(self.mask)
+        print("upsampling mask ...")
+        # upsampling factor in XY plane 
+        mask = np.zeros(self.aff.shape[1:], dtype=self.mask.dtype)
+        for offset in np.ndindex((self.xyfactor, self.xyfactor)):
+            mask[:, np.s_[offset[0]::self.xyfactor], np.s_[offset[1]::self.xyfactor]] = self.mask 
+
+        assert mask.shape == self.aff.shape[1:]
+        assert np.any(self.mask)
+        np.multiply(self.aff[0,:,:,:], mask, out=self.aff[0,:,:,:])
+        np.multiply(self.aff[1,:,:,:], mask, out=self.aff[1,:,:,:])
+        np.multiply(self.aff[2,:,:,:], mask, out=self.aff[2,:,:,:])
+        assert np.any(self.aff)
+
+    def _read_image(self):
+        self.vol = CloudVolume(self.image_layer_path, bounded=False, fill_missing=False,
+                               progress=True, mip=self.image_mip, parallel=True)
+        output_slices = self.output_bbox.to_slices()
+        self.input_slices = tuple(slice(s.start - m, s.stop + m) for s, m in
+                                  zip(output_slices, self.cropping_margin_size))
+        # always reverse the indexes since cloudvolume use x,y,z indexing
+        self.image = self.vol[self.input_slices[::-1]]
+        # the cutout is fortran ordered, so need to transpose and make it C order
+        self.image = np.transpose(self.image)
+        self.image = np.ascontiguousarray(self.image)
+        assert self.image.shape[0] == 1
+        self.image = np.squeeze(self.image, axis=0)
+
+    def _inference(self):
+        # prepare for inference
+        from chunkflow.block_inference_engine import BlockInferenceEngine
+        from chunkflow.frameworks.pznet_patch_inference_engine import PZNetPatchInferenceEngine
+        patch_engine = PZNetPatchInferenceEngine(self.convnet_path)
+        self.block_inference_engine = BlockInferenceEngine(
+            patch_inference_engine=patch_engine,
+            patch_size=self.patch_size,
+            overlap=self.patch_overlap,
+            output_key=self.output_key,
+            output_channels=self.num_output_channels)
+
+
+        # inference engine input is a OffsetArray rather than normal numpy array
+        # it is actually a numpy array with global offset
+        from chunkflow.offset_array import OffsetArray
+
+        input_offset = tuple(s.start for s in self.input_slices)
+        input_chunk = OffsetArray(self.image, global_offset=input_offset)
+        self.output = self.block_inference_engine(input_chunk)
+
+    def _crop(self):
+        self.output = self.output[:,
+                                  self.cropping_margin_size[0] : -self.cropping_margin_size[0],
+                                  self.cropping_margin_size[1] : -self.cropping_margin_size[1],
+                                  self.cropping_margin_size[2] : -self.cropping_margin_size[2]]
+
+    def _upload_output(self):
+        vol = CloudVolume(self.output_layer_path, compress='gzip', fill_missing=True,
```

Comment:
Might be a good idea to make the compression type configurable between gzip and fpzip.

Reply:
this was defined in the `info` file, not here?

Reply:
good point.